cpp/src/io/json/reader_impl.cu

```diff
@@ -111,15 +72,176 @@ constexpr size_t calculate_max_row_size(int num_columns = 0) noexcept
+  out_columns.emplace_back(std::move(result.second[0].results[0]));  // offsets
+  out_columns.emplace_back(std::move(result.second[1].results[0]));  // lengths
+  out_columns.emplace_back(std::move(result.first->release()[0]));   // hashes
+  return std::make_unique<table>(std::move(out_columns));
+}
+
+/**
+ * @brief Initializes the (key hash -> column index) hash map.
+ */
+col_map_ptr_type create_col_names_hash_map(column_view column_name_hashes, cudaStream_t stream)
+{
+  auto key_col_map{col_map_type::create(column_name_hashes.size())};
+  auto const column_data = column_name_hashes.data<uint32_t>();
+  thrust::for_each_n(rmm::exec_policy(stream)->on(stream),
+                     thrust::make_counting_iterator<size_type>(0),
+                     column_name_hashes.size(),
+                     [map = *key_col_map, column_data] __device__(size_type idx) mutable {
+                       map.insert(thrust::make_pair(column_data[idx], idx));
+                     });
+  return std::move(key_col_map);
+}
+
+/**
+ * @brief Create a table whose columns contain the information on JSON objects' keys.
+ *
+ * The columns contain name offsets in the file, name lengths and name hashes, respectively.
+ *
+ * @param[in] data Input JSON device data
+ * @param[in] record_starts Device array of row start locations in the input buffer
+ * @param[in] opts Parsing options (e.g. delimiter and quotation character)
+ * @param[in] stream CUDA stream used for device memory operations and kernel launches
+ *
+ * @return std::unique_ptr<table> cudf table with three columns (offsets, lenghts, hashes)
+ */
+std::unique_ptr<table> create_json_keys_info_table(
+  rmm::device_buffer const &data,
+  rmm::device_vector<uint64_t> const &record_starts,
+  const ParseOptions &opts,
+  cudaStream_t stream)
+{
+  // count keys
+  rmm::device_scalar<unsigned long long int> key_counter(0, stream);
+  auto const data_ptr = static_cast<const char *>(data.data());
+  cudf::io::json::gpu::collect_keys_info(data_ptr,
+                                         data.size(),
+                                         opts,
+                                         record_starts.data().get(),
+                                         record_starts.size(),
+                                         key_counter.data(),
+                                         nullptr,
+                                         stream);
+
+  // Allocate columns to store hash values, lengths, offsets
+  auto const num_keys = key_counter.value();
+  std::vector<std::unique_ptr<column>> info_columns;
+  info_columns.emplace_back(make_numeric_column(data_type(type_id::UINT64), num_keys));
+  info_columns.emplace_back(make_numeric_column(data_type(type_id::UINT16), num_keys));
+  info_columns.emplace_back(make_numeric_column(data_type(type_id::UINT32), num_keys));
+  auto info_table = std::make_unique<table>(std::move(info_columns));
+  rmm::device_scalar<mutable_table_device_view> info_table_mdv(
```

Comment:
This seems weird to me, but I could not come up with a better way to create an optional kernel parameter. Reviewers, is this a sensible approach, do we have better options?

Reply:
Can you elaborate on what this piece of code is doing?

Reply:
@codereport  I'm passing a pointer to the output table here. When the pointer is null, the kernel only counts the object keys. When the table parameter is set, the table is filled with info.
However, passing a table view as an optional parameter here is more complicated than I expected. Just wondering if there is a better way that I missed.

Reply:
C++17 has `std::optional` which sounds like exactly what we want here - but with C++14 using `nullptr` for the "optional-ness" seems fine to me.